neutron/neutron.go

```diff
@@ -0,0 +1,381 @@
+	if err != nil {
+		return nil, errors.Newf(err, "failed to list floating ips")
+	}
+	return resp.FloatingIPV2s, nil
+}
+
+// GetFloatingIPV2 lists details of the floating IP address associated with specified id.
+func (c *Client) GetFloatingIPV2(ipId string) (*FloatingIPV2, error) {
+	var resp struct {
+		FloatingIPV2 FloatingIPV2 `json:"floatingip"`
+	}
+
+	url := fmt.Sprintf("%s/%s", ApiFloatingIPsV2, ipId)
+	requestData := goosehttp.RequestData{RespValue: &resp}
+	err := c.client.SendRequest(client.GET, "network", "v2.0", url, &requestData)
+	if err != nil {
+		return nil, errors.Newf(err, "failed to get floating ip %s details", ipId)
+	}
+	return &resp.FloatingIPV2, nil
+}
+
+// AllocateFloatingIPV2 allocates a new floating IP address in the given external network.
+func (c *Client) AllocateFloatingIPV2(floatingNetworkId string) (*FloatingIPV2, error) {
+	var req struct {
+		FloatingIPV2 struct {
+			FloatingNetworkId string `json:"floating_network_id"`
+		} `json:"floatingip"`
+	}
+	req.FloatingIPV2.FloatingNetworkId = floatingNetworkId
+	var resp struct {
+		FloatingIPV2 FloatingIPV2 `json:"floatingip"`
+	}
+	requestData := goosehttp.RequestData{
+		ReqValue:       req,
+		RespValue:      &resp,
+		ExpectedStatus: []int{http.StatusCreated},
+	}
+	err := c.client.SendRequest(client.POST, "network", "v2.0", ApiFloatingIPsV2, &requestData)
+	if err != nil {
+		return nil, errors.Newf(err, "failed to allocate a floating ip")
+	}
+	return &resp.FloatingIPV2, nil
+}
+
+// DeleteFloatingIPV2 deallocates the floating IP address associated with the specified id.
+func (c *Client) DeleteFloatingIPV2(ipId string) error {
+	url := fmt.Sprintf("%s/%s", ApiFloatingIPsV2, ipId)
+	requestData := goosehttp.RequestData{ExpectedStatus: []int{http.StatusNoContent}}
+	err := c.client.SendRequest(client.DELETE, "network", "v2.0", url, &requestData)
+	if err != nil {
+		err = errors.Newf(err, "failed to delete floating ip %s details", ipId)
+	}
+	return err
+}
+
+// SecurityGroupRuleV2 describes a rule of a security group. There are 2
+// basic rule types: ingress and egress rules (see RuleInfo struct).
+type SecurityGroupRuleV2 struct {
+	PortRangeMax   *int    `json:"port_range_max"` // Can be nil
+	PortRangeMin   *int    `json:"port_range_min"` // Can be nil
+	IPProtocol     *string `json:"protocol"`       // Can be nil, must be defined if PortRange is used
+	ParentGroupId  string  `json:"security_group_id"`
+	RemoteIPPrefix string  `json:"remote_ip_prefix"`
+	RemoteGroupID  string  `json:"remote_group_id"`
+	EthernetType   string  `json:"ethertype"`
+	Direction      string  `json:"direction"` // Required
+	Id             string  `json:",omitempty"`
+	TenantId       string  `json:"tenant_id,omitempty"`
+}
+
+// SecurityGroupV2 describes a single security group in OpenStack.
+type SecurityGroupV2 struct {
+	Rules       []SecurityGroupRuleV2 `json:"security_group_rules"`
+	TenantId    string                `json:"tenant_id"`
+	Id          string
```

Comment:
I'd give json lowercased names for these as well so they reserialise lowercased.